File: generated-code/cpp/codecraft/FileReadWrite/model/MoveAction.cpp
#include "MoveAction.hpp"

namespace model {

MoveAction::MoveAction(Vec2Int target, bool findClosestPosition, bool breakThrough) : target(target), findClosestPosition(findClosestPosition), breakThrough(breakThrough) { }

// Read MoveAction from input stream
MoveAction MoveAction::readFrom(InputStream& stream) {
    Vec2Int target = Vec2Int::readFrom(stream);
    bool findClosestPosition = stream.readBool();
    bool breakThrough = stream.readBool();
    return MoveAction(target, findClosestPosition, breakThrough);
}

// Write MoveAction to output stream
void MoveAction::writeTo(OutputStream& stream) const {
    target.writeTo(stream);
    stream.write(findClosestPosition);
    stream.write(breakThrough);
}

// Get string representation of MoveAction
std::string MoveAction::toString() const {
    std::stringstream ss;
    ss << "MoveAction { ";
    ss << "target: ";
    ss << target.toString();
    ss << ", ";
    ss << "findClosestPosition: ";
    ss << findClosestPosition;
    ss << ", ";
    ss << "breakThrough: ";
    ss << breakThrough;
    ss << " }";
    return ss.str();
}

bool MoveAction::operator ==(const MoveAction& other) const {
    return target == other.target && findClosestPosition == other.findClosestPosition && breakThrough == other.breakThrough;
}

}

size_t std::hash<model::MoveAction>::operator ()(const model::MoveAction& value) const {
    size_t result = 0;
    result ^= std::hash<Vec2Int>{}(value.target) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<bool>{}(value.findClosestPosition) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<bool>{}(value.breakThrough) + 0x9e3779b9 + (result << 6) + (result >> 2);
    return result;
}

File: generated-code/go/codecraft/FileReadWrite/model/move_action.go
package model

import "fmt"
import "io"
import . "trans_gen_test/common"
import . "trans_gen_test/stream"

// Move action
type MoveAction struct {
    // Target position
    Target Vec2Int32
    // Whether to try find closest position, if path to target is not found
    FindClosestPosition bool
    // Whether to destroy other entities on the way
    BreakThrough bool
}

func NewMoveAction(target Vec2Int32, findClosestPosition bool, breakThrough bool) MoveAction {
    return MoveAction {
        Target: target,
        FindClosestPosition: findClosestPosition,
        BreakThrough: breakThrough,
    }
}

// Read MoveAction from reader
func ReadMoveAction(reader io.Reader) MoveAction {
    var target Vec2Int32
    target = ReadVec2Int32(reader)
    var findClosestPosition bool
    findClosestPosition = ReadBool(reader)
    var breakThrough bool
    breakThrough = ReadBool(reader)
    return MoveAction {
        Target: target,
        FindClosestPosition: findClosestPosition,
        BreakThrough: breakThrough,
    }
}

// Write MoveAction to writer
func (moveAction MoveAction) Write(writer io.Writer) {
    target := moveAction.Target
    target.Write(writer)
    findClosestPosition := moveAction.FindClosestPosition
    WriteBool(writer, findClosestPosition)
    breakThrough := moveAction.BreakThrough
    WriteBool(writer, breakThrough)
}

// Get string representation of MoveAction
func (moveAction MoveAction) String() string {
    stringResult := "{ "
    stringResult += "Target: "
    target := moveAction.Target
    stringResult += target.String()
    stringResult += ", "
    stringResult += "FindClosestPosition: "
    findClosestPosition := moveAction.FindClosestPosition
    stringResult += fmt.Sprint(findClosestPosition)
    stringResult += ", "
    stringResult += "BreakThrough: "
    breakThrough := moveAction.BreakThrough
    stringResult += fmt.Sprint(breakThrough)
    stringResult += " }"
    return stringResult
}